pennylane/transforms/qfunc_transforms.py

```diff
@@ -0,0 +1,306 @@
+    quantum function, without any of the operations being queued by
+    any existing queuing contexts:
+
+    >>> with qml.tape.QuantumTape() as active_tape:
+    ...     qml.RY(1.0, wires=0)
+    ...     tape = make_tape(qfunc)(0.5)
+    >>> tape.operations
+    [Hadamard(wires=[0]), CNOT(wires=[0, 1]), RX(0.5, wires=[0])]
+
+    Note that the currently recording tape did not queue any of these quantum operations:
+
+    >>> active_tape.operations
+    [RY(1.0, wires=[0])]
+    """
+
+    def wrapper(*args, **kwargs):
+        active_tape = qml.tape.get_active_tape()
+
+        if active_tape is not None:
+            with active_tape.stop_recording(), active_tape.__class__() as tape:
+                fn(*args, **kwargs)
+        else:
+            with qml.tape.QuantumTape() as tape:
+                fn(*args, **kwargs)
+        return tape
+
+    return wrapper
+
+
+class NonQueuingTape(qml.queuing.AnnotatedQueue):
+    """Mixin class that creates a tape that does not queue
+    itself to the current queuing context."""
+
+    def _process_queue(self):
+        super()._process_queue()
+
+        for obj, info in self._queue.items():
+            qml.queuing.QueuingContext.append(obj, **info)
+
+        qml.queuing.QueuingContext.remove(self)
+
+
+class single_tape_transform:
+    """For registering a tape transform that takes a tape and outputs a single new tape.
+
+    Examples of such transforms include circuit compilation.
+
+    Args:
+        transform_fn (function): The function to register as the single tape transform.
+            It can have an arbitrary number of arguments, but the first argument
+            **must** be the input tape.
+
+    **Example**
+
+    A valid single tape transform is a quantum function that satisfies the following:
+
+    - The first argument must be an input tape
+
+    - Depending on the structure of this input tape, various quantum operations, functions,
+      and templates may be called.
+
+    - Any internal classical processing should use the ``qml.math`` module to ensure
+      the transform is differentiable.
+
+    - There is no return statement.
+
+    For example:
+
+    .. code-block:: python
+
+        @qml.single_tape_transform
+        def my_transform(tape, x, y):
+            # loop through all operations on the input tape
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    wires = op.wires
+                    param = op.parameters[0]
+
+                    qml.RX(x * qml.math.abs(param), wires=wires[1])
+                    qml.RY(y * qml.math.abs(param), wires=wires[1])
+                    qml.CZ(wires=[wires[1], wires[0]])
+                else:
+                    op.queue()
+
+    This transform iterates through the input tape, and replaces any :class:`~.CRX` operation with
+    two single qubit rotations and a :class:`~.CZ` operation. These newly queued operations will

```

Comment:
Is this a decomposition of `CRX`? Any particular reason someone would be doing this?

Reply:
No, this isn't a decomposition, it is completely arbitrary :) The only reason is to serve as an example for now